src/shadowbox/server/server_access_key.spec.ts

```diff
@@ -92,8 +94,102 @@ describe('ServerAccessKeyRepository', () => {
     });
   });
 
+  it('Creates keys at the right port by construction', async (done) => {
+    const portProvider = new PortProvider();
+    const port = await portProvider.reserveNewPort();
+    const repo = new RepoBuilder().port(port).build();
+    const key = await repo.createNewAccessKey();
+    expect(key.proxyParams.portNumber).toEqual(port);
+    done();
+  });
+
+  it('setDefaultPort changes default port for new keys', async (done) => {
+    const portProvider = new PortProvider();
+    const port = await portProvider.reserveNewPort();
+    const repo = new RepoBuilder().build();
+    await repo.setDefaultPort(port);
+    const key = await repo.createNewAccessKey();
+    expect(key.proxyParams.portNumber).toEqual(port);
+    done();
+  });
+
+  it('setDefaultPort maintains ports on existing keys', async (done) => {
+    const portProvider = new PortProvider();
+    const oldPort = await portProvider.reserveNewPort();
+    const repo = new RepoBuilder().port(oldPort).build();
+    const oldKey = await repo.createNewAccessKey();
+
+    const newPort = await portProvider.reserveNewPort();
+    await repo.setDefaultPort(newPort);
+    expect(oldKey.proxyParams.portNumber).toEqual(oldPort);
+    done();
+  });
+
+  it('setDefaultPort rejects invalid port numbers', async (done) => {
+    const repo = new RepoBuilder().build();
+    // jasmine.toThrowError expects a function and makes the code
+    // hard to read.
+    const expectThrow = async (port: number) => {
+      try {
+        await repo.setDefaultPort(port);
+        fail(`setDefaultPort should reject invalid port number ${port}.`);
+      } catch (error) {
+        expect(error instanceof errors.InvalidPortNumber).toBeTruthy();
+      }
+    };
+    await expectThrow(0);
+    await expectThrow(-1);
+    await expectThrow(100.1);
+    await expectThrow(65536);
+    done();
+  });
+
+  it('setDefaultPort rejects ports in use', async (done) => {
+    const portProvider = new PortProvider();
+    const port = await portProvider.reserveNewPort();
+    const repo = new RepoBuilder().build();
+    const server = new net.Server();
+    server.listen(port, async () => {
+      try {
+        await repo.setDefaultPort(port);
+        fail(`setDefaultPort should reject already used port ${port}.`);
+      } catch (error) {
+        expect(error instanceof errors.PortInUse);
+      }
+      server.close();
+      done();
+    });
+  });
+
+  it('setDefaultPort accepts ports already used by access keys', async (done) => {
+    const portProvider = new PortProvider();
+    const oldPort = await portProvider.reserveNewPort();
+    const repo = new RepoBuilder().port(oldPort).build();
+    await repo.createNewAccessKey();
+
+    // jasmine.toThrowError expects a function and makes the code
```

Comment:
This won't work because toThrow() wants expect() to take a function, not an expression.  I also don't think we can use the option we discussed offline works either.  I updated the comment to explain.